### ISSUE TITLE
DataVec: Issue loading multi-channel, multi-page TIFF

### ISSUE THREAD

**Issue by AlexDBlack** [eclipse/deeplearning4j#7689]
Image appears to have 2 channels.
```
        NativeImageLoader n = new NativeImageLoader(256,256, 2, BaseImageLoader.MultiPageMode.CHANNELS);
        String s = "C:\\Users\\Alex\\Downloads\\3d.tiff";
        INDArray arr = n.asMatrix(new File(s));
        System.out.println(Arrays.toString(arr.shape()));
```
[3d.zip](https://github.com/deeplearning4j/deeplearning4j/files/3157169/3d.zip)

```
Exception in thread "main" java.lang.IllegalStateException: Cannot assign arrays: arrays must both be scalars, both vectors, or shapes must be equal other than size 1 dimensions. Attempting to do x.assign(y) with x.shape=[1, 256, 256] and y.shape=[1, 2, 256, 256]
	at org.nd4j.base.Preconditions.throwStateEx(Preconditions.java:641)
	at org.nd4j.base.Preconditions.checkState(Preconditions.java:340)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.assign(BaseNDArray.java:1398)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.put(BaseNDArray.java:2378)
	at org.datavec.image.loader.NativeImageLoader.asMatrix(NativeImageLoader.java:856)
	at org.datavec.image.loader.NativeImageLoader.asMatrix(NativeImageLoader.java:246)
	at org.datavec.image.loader.NativeImageLoader.asMatrix(NativeImageLoader.java:237)
	at org.deeplearning4j.examples.Temp.main(Temp.java:16)
```

**Comment by saudet** [comment]
@BioinfoTongLI Could you help us fix this? It should be pretty easy

**Comment by BioinfoTongLI** [comment]
Sure. But since it has been a while, a lot of things have changed. I failed to run datavec tests in the monorepo. Do you have a tuto for this ? I just `mvn clean install` it. Is that ok?

Before starting the modification, I want to confirm sth with you.
1. I see a `.tiff` with shape of (181, 256, 256) channel equals 1. In my opinion, it's z-axis which equals 181. Why are you forcing it to load with 2 channels?
2. I remember that Datavec was not supposed to have this unusual "Z" dimension, isn't it? If it's still the case, which dimension should I put it into ?

**Comment by saudet** [comment]
Thanks for your help! `mvn clean install` in the `datavec` subdirectory with say `-Ptest-nd4j-native` should just work, yes. What error are you getting?

I think the 3d.tiff file above has only 1 channel of 16 bits, but there are TIFF files with multiple pages and more than 1 channel, and those should be loaded properly anyway.

As for the format of the loaded arrays, I think DL4J standardized on NCDHW so we should probably just load them all like that by default and let users permute if needed, @AlexDBlack ?

**Comment by AlexDBlack** [comment]
Unlike DL4J CNN2D layers (which are NCHW only), Conv3d in DL4J and SameDiff support both NCDHW and NDHWC format:
https://github.com/deeplearning4j/deeplearning4j/blob/0ff3ced27d454a527ecf13aba03c7c0d3e4d882a/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Convolution3D.java#L49-L57
NCDHW is the default in DL4J, and yes, it's just a permute that is required to switch to NDHWC.

As far as that 3d.tiff file - DataVec seems to think it's 2 channel for some reason, I'm getting an exception (about not being able to load 2 channels into 1) when I set it to 1 channel in the image loader constructor.

**Comment by BioinfoTongLI** [comment]
Ah, @saudet thank you for pointing it out. I think it's because it's a 16bit-image, so the loader try to load it into two channels. 
I remember that other than 8-bit depth was not supported. So is 16bit officially supported now ?

**Comment by saudet** [comment]
It's always been possible to do 16 or 32 bits, we just need to implement it :)

**Comment by BioinfoTongLI** [comment]
After a quick look, it seems that it's the "cast" in leptonica wrapper that consider 16bit as a 8 bit image with `pixa.n() == 2`

https://github.com/deeplearning4j/deeplearning4j/blob/1342dcd0fa8e7eb30c912fd2235e4f71349aec6a/datavec/datavec-data/datavec-data-image/src/main/java/org/datavec/image/loader/NativeImageLoader.java#L821-L830

this is the wrap function
```public static native PIXA pixaReadMemMultipageTiff(@Cast({"const l_uint8*"}) BytePointer var0, @Cast({"size_t"}) long var1);```

What's your opinion here ?

**Comment by saudet** [comment]
We get the number of channels from `PIX.d` like this:
https://github.com/bytedeco/javacv/blob/master/src/main/java/org/bytedeco/javacv/LeptonicaFrameConverter.java#L111
It's not related to `PIXA.n`.

**Comment by BioinfoTongLI** [comment]
Sorry, my mistake. It's been a while I write no more Java.
Actually, the`pixa.n()` is the total number of images....I can not even understand my code. This is bad

**Comment by BioinfoTongLI** [comment]
This is the method to change.

https://github.com/deeplearning4j/deeplearning4j/blob/1342dcd0fa8e7eb30c912fd2235e4f71349aec6a/datavec/datavec-data/datavec-data-image/src/main/java/org/datavec/image/loader/NativeImageLoader.java#L192-L227

 The cases `pix.d() > 8` are ignored. Everything before was force to 8bit.
So what to do with 16 and 32 bit ? squezze into 8 bit with lossy compression ?

**Comment by saudet** [comment]
No, we can just read them as 16 and 32 bits and load that into NDArray.

**Comment by BioinfoTongLI** [comment]
I added the wrapper for 16 and 32 bits. It still needs some clean-up, but all tests passes now.
About the image data for unit test. Do I have to PR in the testressource repo ?Or maybe you can upload something and I adapt it in my tests?

**Comment by AlexDBlack** [comment]
@BioinfoTongLI yeah, just open a PR on dl4j-test-resources if you want to use that in a unit test.

**Comment by agibsonccc** [comment]
Closing — references a removed/rewritten subsystem (DataVec, NLP/Word2Vec, Model Zoo, Spark, UI, or Arbiter). These have been fundamentally rewritten or removed in the upcoming release. Please open a new issue if you experience similar problems with the new version.
